### PR TITLE
feat: add ORPO trainer support

### PR DESCRIPTION
**Pull request by eghuzefa**
## Summary

This PR implements **ORPO (Odds Ratio Preference Optimization)** as a memory-efficient alternative to DPO for preference tuning. ORPO achieves approximately **50% memory savings** by eliminating the need for a reference model during training.

## Motivation

ORPO provides the same preference learning capabilities as DPO while requiring only a single model forward pass, making it ideal for resource-constrained environments and large model training.

**Paper**: https://arxiv.org/abs/2403.07691

### Core Implementation
- **`tunix/rl/orpo/orpo_trainer.py`**: Complete ORPO trainer implementation (516 lines)
  - `ORPOTrainer` class inheriting from `PeftTrainer`
  - `ORPOTrainingConfig` dataclass with `lambda_orpo` hyperparameter
  - `orpo_loss_fn` implementing odds ratio-based preference optimization
  - Support for both tokenized and string inputs
  
- **`tunix/rl/orpo/__init__.py`**: Module exports

### Testing
- **`tests/rl/orpo/orpo_trainer_test.py`**: Comprehensive test suite
  - 6 tests covering training loop, loss computation, input handling
  - All tests passing ✅
  - End-to-end training verified

### Integration
- **`tunix/__init__.py`**: Added ORPO exports to main package API

## Key Features

- ✅ **No reference model required** (key difference from DPO)
- ✅ **Memory efficient**: ~50% memory savings vs DPO
- ✅ **API compatible**: Works with existing `PeftTrainer` infrastructure
- ✅ **Full metrics logging**: `rewards/chosen`, `rewards/rejected`, `rewards/margin`, `rewards/accuracy`, `odds_ratio`
- ✅ **Flexible input formats**: Supports both tokenized and string inputs

### CODE REVIEW

**Comment by wang2yn84** [comment]
Hi [eghuzefa](https://github.com/eghuzefa), thank you for your PR. Do you happen to formatted everything? It's hard to review the actual change right now. Can you format the files that you changed?

**Comment by eghuzefa** [comment]
> Hi [eghuzefa](https://github.com/eghuzefa), thank you for your PR. Do you happen to formatted everything? It's hard to review the actual change right now. Can you format the files that you changed?

Hey, yes I did, two files were primarily added 
tunix/rl/orpo/orpo_trainer.py
tests/rl/orpo/orpo_trainer_test.py

and the following file was changed for adding the trainer export
tunix/__init__.py

**Comment by eghuzefa** [comment]
I will go through it again, fix all these conflicts, and reformat my changes, by tomorrow.

**Comment by wang2yn84** [comment]
Hi [eghuzefa](https://github.com/eghuzefa), please let me know when this is PR ready for review. Right now still have formatting issues.

**Comment by eghuzefa** [comment]
> Hi [eghuzefa](https://github.com/eghuzefa), please let me know when this is PR ready for review. Right now still have formatting issues.

Thanks for the review! I had rebased  2 days ago on the latest upstream/main and today made the following changes to match the codebase patterns: I updated ORPO to be compatible with recent upstream API changes: common.get_per_token_logps() now returns a single value instead of a tuple, and I renamed the _prepare_inputs parameter from training_input to input_data to match the base class signature. I also removed the __init__.py files from both tunix/rl/orpo/ and tests/rl/orpo/ to exactly match the structure used by DPO and GRPO (both export directly from tunix/__init__.py without intermediate __init__.py files). All 6 tests pass, pyink formatting is clean, and pylint gives 9.78/10 with only minor non-blocking docstring warnings. Could you please clarify what specific formatting issues you're seeing? I've run pre-commit hooks and everything passes. Are there specific files or lines that need attention, or perhaps a different formatting tool I should use?

**Comment by wang2yn84** [comment]
Hi @eghuzefa , I'm seeing 42 files are changes in this PR, I believe that's because you run pre-commit manually on the whole repo? It's hard to focus on the actual change you have. Can you remove those format only changes? @sizhit2 , can you help take a look at the orpo_trainer?

**Comment by eghuzefa** [comment]
> Hi @eghuzefa , I'm seeing 42 files are changes in this PR, I believe that's because you run pre-commit manually on the whole repo? It's hard to focus on the actual change you have. Can you remove those format only changes? @sizhit2 , can you help take a look at the orpo_trainer?

Hey @wang2yn84, yeah you're right — I ran pre-commit on the whole repo. I’ll revert the formatting-only changes and keep just the relevant ones.

**Comment by eghuzefa** [comment]
> > Hi @eghuzefa , I'm seeing 42 files are changes in this PR, I believe that's because you run pre-commit manually on the whole repo? It's hard to focus on the actual change you have. Can you remove those format only changes? @sizhit2 , can you help take a look at the orpo_trainer?
> 
> Hey @wang2yn84, yeah you're right — I ran pre-commit on the whole repo. I’ll revert the formatting-only changes and keep just the relevant ones.

Hi @wang2yn84, I have fixed it now, kindly check.

**Comment by wang2yn84** [comment]
Thank you @eghuzefa, it's much better now! Couple of questions here: 1. ORPO is a variation of DPO, can you move to sft/ ? 2. We are refactoring the logics now to abstract the core algorithm. Before that, can you merge the algorithm directly to dpo? Most of the code added should be similar to DPO except some additional configs and the loss function.

**Comment by eghuzefa** [comment]
> Thank you @eghuzefa, it's much better now! Couple of questions here: 1. ORPO is a variation of DPO, can you move to sft/ ? 2. We are refactoring the logics now to abstract the core algorithm. Before that, can you merge the algorithm directly to dpo? Most of the code added should be similar to DPO except some additional configs and the loss function.

Done! I've merged ORPO into DPO in sft/. Both algorithms now share the same trainer with algorithm="orpo" config and no reference model. I kept the _dpo_ folder name to minimize changes - happy to rename it to something more generic like _preference_ or _po_ if you prefer, but wanted to avoid unnecessary refactoring for now. All tests passing ✓

File: tunix/sft/dpo/dpo_trainer.py [hunk 1]
```diff
@@ -262,9 +296,10 @@ def _prepare_inputs(
     positions = common.build_positions_from_mask(mask)
 
     # Compute the log probabilities for the chosen and rejected tokens.
+    # For ORPO, we don't use a reference model.
     ref_chosen_logps = None
     ref_rejected_logps = None
-    if self._ref_model_exists:
+    if self._ref_model_exists and self.algorithm == "dpo":
```
Comment (wang2yn84) [inline tunix/sft/dpo/dpo_trainer.py]: It's still safe to rely on self._ref_model_exists here, because in ORPO we don't have ref_model. Let's revert it here?

File: tunix/sft/dpo/dpo_trainer.py [hunk 2]
```diff
@@ -328,33 +377,83 @@ def dpo_loss_fn(
...
-      "log_probs/chosen": chosen_logps.mean(),
-      "log_probs/rejected": rejected_logps.mean(),
-  }
+    # L_OR: Odds ratio preference loss
+    # Uses log odds ratio between chosen and rejected responses
+    log_odds = chosen_logps - rejected_logps
 
-  return losses.mean(), aux
+    # Apply label smoothing to odds ratio loss
+    or_loss = -(
```
Comment (wang2yn84) [inline tunix/sft/dpo/dpo_trainer.py:403]: Any reason why we have the label smoothing here? Is it supposed to be log_odds_chosen = chosen_logps - jnp.log1p(-jnp.exp(chosen_logps)) ? I see huggingface have similar impl: https://github.com/huggingface/trl/blob/02a34777c36173ea53c8ce9979db50f13c1aaac8/trl/trainer/orpo_trainer.py#L675C1-L677C10

File: tunix/sft/dpo/dpo_trainer.py [hunk 3]
```diff
@@ -328,33 +377,83 @@ def dpo_loss_fn(
...
+        + jax.nn.log_sigmoid(-log_odds) * label_smoothing
+    )
+
+    # Combined ORPO loss: L_ORPO = L_SFT + λ * L_OR
+    total_loss = sft_loss + lambda_orpo * or_loss
+
+    # Compute rewards for logging
+    # Use sigmoid of log odds to get meaningful 0-1 preference probability
+    preference_prob = jax.nn.sigmoid(log_odds)
+    chosen_rewards = preference_prob
```
Comment (wang2yn84) [inline tunix/sft/dpo/dpo_trainer.py]: The rewards calculation is also different from TRL, do you have some other reference for this?

File: tunix/sft/dpo/dpo_trainer.py (hunk 2 above)
Comment (eghuzefa) [reply tunix/sft/dpo/dpo_trainer.py]: Initially used simplified log probability ratio (log(P_chosen/P_rejected)) instead of true odds ratio to avoid numerical issues with exp() operations. Now fixed to proper odds ratio formula log(P/(1-P)) using log_sigmoid for numerical stability, matching both the paper and HuggingFace implementation. No NaN issues

File: tunix/sft/dpo/dpo_trainer.py (hunk 1 above)
Comment (eghuzefa) [reply tunix/sft/dpo/dpo_trainer.py]: Done!

File: tunix/sft/dpo/dpo_trainer.py (hunk 3 above)
Comment (eghuzefa) [reply tunix/sft/dpo/dpo_trainer.py]: I am working on these issues, give me some time.

File: tunix/sft/dpo/dpo_trainer.py (hunk 2 above)
Comment (eghuzefa) [reply tunix/sft/dpo/dpo_trainer.py:403]: I am working on these issues, give me some time.

Comment (eghuzefa) [reply tunix/sft/dpo/dpo_trainer.py:403]: Label smoothing is standard regularization (not ORPO-specific, DPO has it too). For the odds calculation - yes, now fixed! We use the exact HuggingFace TRL formula you referenced. Do let me know if you see any more issue on the code.

File: tunix/sft/dpo/dpo_trainer.py (hunk 3 above)
Comment (eghuzefa) [reply tunix/sft/dpo/dpo_trainer.py]: Good catch! Fixed to match TRL exactly: lambda_orpo * logps (like TRL's beta * logps). Note on paper vs implementation: The paper doesn't explicitly define reward calculation for logging - this is a TRL convention we're following for consistency with the ecosystem. Our lambda_orpo serves the same role as TRL's beta parameter. Please let me know if there are any other issues you see in the code.

File: tunix/sft/dpo/dpo_trainer.py (hunk 2 above)
Comment (eghuzefa) [reply tunix/sft/dpo/dpo_trainer.py:403]: Note: log1p(-exp(x)) requires x < 0 to avoid NaN. This works when log probs are averaged per token, but may produce NaN for summed log probs (though highly unlikely). TRL uses summed log probs and relies on them being negative.

File: tunix/sft/dpo/dpo_trainer.py (hunk 3 above)
Comment (eghuzefa) [reply tunix/sft/dpo/dpo_trainer.py]: Note: For reward calculation, TRL uses 'beta' as the scaling factor, but we use lambda_orpo to distinguish it from DPO's beta (KL weight), since we have a unified trainer for both algorithms while TRL uses separate classes.

File: tunix/sft/dpo/dpo_trainer.py (hunk 2 above)
Comment (wang2yn84) [reply tunix/sft/dpo/dpo_trainer.py:403]: Since probs is (0, 1) so log probs are strictly < 0, no matter if it's averaged or summed. Wdyt?

File: tunix/sft/dpo/dpo_trainer.py (hunk 3 above)
Comment (wang2yn84) [reply tunix/sft/dpo/dpo_trainer.py]: Yea I get this part. I think it looks good now. I like this naming.

**Comment by wang2yn84** [comment]
Thank you so much for the support! Looks great!

File: tunix/sft/dpo/dpo_trainer.py (hunk 2 above)
Comment (eghuzefa) [reply tunix/sft/dpo/dpo_trainer.py:403]: Yes, that makes sense!

**Comment by wang2yn84** [comment]
Do you have a gmail account? Can we chat a bit more about your contribution?

**Comment by wang2yn84** [comment]
One last thing, can you squash your commits into 1?

**Comment by eghuzefa** [comment]
> Do you have a gmail account? Can we chat a bit more about your contribution?

hey, sure, here it is, m.huzefa1993@gmail.com

**Comment by eghuzefa** [comment]
Apologies for the noise!
While squashing my commits into a single clean commit, I made a mistake during the git reset process that caused the PR to auto-close. The PR has been reopened with all the ORPO implementation changes intact and consolidated into one commit.
Sorry for any confusion this caused!

**Comment by wang2yn84** [comment]
> Apologies for the noise! While squashing my commits into a single clean commit, I made a mistake during the git reset process that caused the PR to auto-close. The PR has been reopened with all the ORPO implementation changes intact and consolidated into one commit. Sorry for any confusion this caused!

Oh no worry, I saw some github workflow test failure last time, need to fix them before we can merge.

**Comment by wang2yn84** [comment]
> > Do you have a gmail account? Can we chat a bit more about your contribution?
> 
> hey, sure, here it is, [m.huzefa1993@gmail.com](mailto:m.huzefa1993@gmail.com)

Are you available on Google chat?
